### PR TITLE
[C#] regular expressions new exercise

### PR DESCRIPTION
**Pull request by mikedamay**
This follows on from #1727 (as a result of previous problems with my forked repo).

#1727 was dealt with fully except for an issue on tasks 3 & 5 which is still open.  The current code here has the variation in wording referred to in my penultimate comment on #1727

### CODE REVIEW

File: languages/csharp/exercises/concept/regular-expressions/.docs/instructions.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+It is important to find any passwords included in a file. These will be dealt with automatically, but the team needs to know how about passwords occurred in quoted text so that they can be examined manually.
+
+Implement the `LogParser.CountQuotedPasswords()`method to provide an indication of the likely scale of the manual exercise.
+
+The literal string "password" may be in upper or lower case or any combination.
+
+Lines passed to the routine may or may not be valid as defined in task 1. We process them in the same way, whether or not they are valid.
+
+```csharp
+string[] lines =
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/regular-expressions/.docs/instructions.md:55]: This example is using an array but the method itself uses a string.

**Comment by mikedamay** [comment]
closed to finalize - will reopen shortly

**Comment by mikedamay** [comment]
All files complete and proofed

File: languages/csharp/config.json [hunk 2]
```diff
@@ -24,6 +24,12 @@
         "concepts": ["booleans"],
         "prerequisites": ["basics"]
       },
+      {
+        "slug": "casting",
```
Comment (ErikSchierboom) [inline languages/csharp/config.json:28]: I think this should be part of another PR?

File: languages/csharp/config.json [hunk 3]
```diff
@@ -178,6 +196,18 @@
         "concepts": ["strings"],
         "prerequisites": ["basics"]
       },
+      {
+        "slug": "structs",
```
Comment (ErikSchierboom) [inline languages/csharp/config.json:200]: I think this should be part of another PR?

File: languages/csharp/config.json [hunk 4]
```diff
@@ -166,6 +172,18 @@
...
+      {
+        "slug": "regular-expressions",
+        "uuid": "c63bb75d-51bd-4d5d-b37d-88e53ee0da69",
+        "concepts": ["regular-expressions"],
+        "prerequisites": [
+          "arrays",
+          "for-loops",
+          "strings",
+          "string-interpolation",
+          "verbatim-strings"
```
Comment (ErikSchierboom) [inline languages/csharp/config.json:184]: We currently don't have separate `string-interpolation` and `verbatim-strings` concepts being unlocked, but we could add these to the `strings` exercise.

File: languages/csharp/exercises/concept/regular-expressions/.docs/hints.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+- [regular expressions][regular-expressions] documentation describes regexes and the flavour built into the .NET libraries.
+- [Regex][regex] documentation describing the built-in library class.
+
+### 1. Identify garbled log lines
+
+See [this discussion][ismatch] of matching.
+
+### 2. Split the log line
+
+[This][split] article shows an approach to splitting lines.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/regular-expressions/.docs/hints.md:12]:
```suggestion
- [This][split] article shows an approach to splitting lines.
```

File: languages/csharp/exercises/concept/regular-expressions/.docs/hints.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
+## General
+
+- [regular expressions][regular-expressions] documentation describes regexes and the flavour built into the .NET libraries.
+- [Regex][regex] documentation describing the built-in library class.
+
+### 1. Identify garbled log lines
+
+See [this discussion][ismatch] of matching.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/regular-expressions/.docs/hints.md:8]:
```suggestion
- See [this discussion][ismatch] of matching.
```

File: languages/csharp/exercises/concept/regular-expressions/.docs/hints.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+### 2. Split the log line
+
+[This][split] article shows an approach to splitting lines.
+
+### 3. Count the number of lines containing a password
+
+See [this discussion][multiple-matches] of multiple matches and [this][multi-line] on multi-line searches.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/regular-expressions/.docs/hints.md:16]:
```suggestion
- See [this discussion][multiple-matches] of multiple matches and [this][multi-line] on multi-line searches.
```

File: languages/csharp/exercises/concept/regular-expressions/.docs/hints.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+### 4. Remove artifacts from log
+
+[This article][replace] shows how text can be replaced.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/regular-expressions/.docs/hints.md:20]:
```suggestion
- [This article][replace] shows how text can be replaced.
```

File: languages/csharp/exercises/concept/regular-expressions/.docs/hints.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+### 5. List lines with extremely weak passwords so the guilty can be punished
+
+Capture groups are discussed [here][capture-groups].
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/regular-expressions/.docs/hints.md:24]:
```suggestion
- Capture groups are discussed [here][capture-groups].
```

File: languages/csharp/exercises/concept/regular-expressions/.meta/Example.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        return matches.Count;
+    }
+
+    public string RemoveEndOfLineText(string line)
+    {
+        string pattern = @"end-of-Line\d+";
+
+        string str = Regex.Replace(line, pattern, string.Empty,
+            RegexOptions.IgnoreCase);
+        return str;
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/regular-expressions/.meta/Example.cs:31]:
```suggestion
        return Regex.Replace(line, pattern, string.Empty,
            RegexOptions.IgnoreCase);
```

File: languages/csharp/exercises/concept/regular-expressions/RegularExpressionsTests.cs [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        string[] expected =
+        {
+            "passWordaa: [INF] passWordaa",
+            "--------: passWord mysecret",
+            "--------: [INF] password KeyToTheCastle for nobody",
+            "password123: [INF] password password123 for everybody"
+        };
+        Assert.Equal(expected, lp.ListLinesWithPasswords(lines));
+    }
+
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/regular-expressions/RegularExpressionsTests.cs:77]:
```suggestion
```

File: languages/csharp/config.json (hunk 2 above)
Comment (mikedamay) [reply languages/csharp/config.json:28]: `casting` is in master but was not in the base of this branch.  Won't it merge as is or do I have to do  something with my branch.  Last time I did anything like that it was a disaster and the PR says This Branch has no Conflicts.

File: languages/csharp/config.json (hunk 3 above)
Comment (mikedamay) [reply languages/csharp/config.json:200]: as for `casting`

File: languages/csharp/config.json (hunk 4 above)
Comment (mikedamay) [reply languages/csharp/config.json:184]: Both are concepts in `string-formatting`.

Comment (ErikSchierboom) [reply languages/csharp/config.json:184]: Brilliant. Thanks

File: languages/csharp/config.json (hunk 2 above)
Comment (ErikSchierboom) [reply languages/csharp/config.json:28]: Ah, I didn't realize it was in master. Just keep it in.

File: languages/csharp/config.json (hunk 3 above)
Comment (ErikSchierboom) [reply languages/csharp/config.json:200]: Okay. Keep it in.

File: languages/csharp/exercises/concept/regular-expressions/.docs/instructions.md (hunk 1 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/regular-expressions/.docs/instructions.md]: done
